File: Sources/TapeNexus/MediaTools.swift
import Foundation

/// Post-download tools that run the bundled ffmpeg on a Library file: extract
/// audio, transcode to MP4, and trim a clip. ffmpeg is the same binary yt-dlp
/// uses for merging bestvideo+bestaudio, so no new dependency is shipped — just
/// `state.yt.ffmpegLocation`. All work runs off the main thread; the completion
/// is delivered on main.
enum MediaTools {
    /// Run `ffmpeg <args>` on a background queue. `args` is the full argument
    /// list (including `-i input` and the output path). Completion gets the
    /// output path on success, nil on failure.
    static func run(ffmpeg: URL, args: [String], completion: @escaping (Bool) -> Void) {
        DispatchQueue.global(qos: .userInitiated).async {
            let p = Process()
            p.executableURL = ffmpeg
            p.arguments = ["-y"] + args
            p.standardOutput = FileHandle()
            p.standardError = FileHandle()
            do {
                try p.run()
                p.waitUntilExit()
                DispatchQueue.main.async { completion(p.terminationStatus == 0) }
            } catch {
                DispatchQueue.main.async { completion(false) }
            }
        }
    }

    /// A non-clobbering output path beside the input: `<name><suffix>.<ext>`,
    /// `-2`, `-3`, … as needed.
    static func outputURL(for input: URL, suffix: String, ext: String) -> URL {
        let dir = input.deletingLastPathComponent()
        let base = input.deletingPathExtension().lastPathComponent
        var candidate = dir.appendingPathComponent("\(base)\(suffix).\(ext)")
        var i = 2
        while FileManager.default.fileExists(atPath: candidate.path) {
            candidate = dir.appendingPathComponent("\(base)\(suffix)-\(i).\(ext)")
            i += 1
        }
        return candidate
    }

    // MARK: Recipes — each returns (args, outputURL) for a given input.

    static func extractAudioMP3(_ input: URL) -> (args: [String], out: URL) {
        let out = outputURL(for: input, suffix: "-audio", ext: "mp3")
        return (["-i", input.path, "-vn", "-c:a", "libmp3lame", "-q:a", "2", out.path], out)
    }

    static func extractAudioAAC(_ input: URL) -> (args: [String], out: URL) {
        let out = outputURL(for: input, suffix: "-audio", ext: "m4a")
        return (["-i", input.path, "-vn", "-c:a", "aac", "-b:a", "192k", out.path], out)
    }

    static func transcodeMP4(_ input: URL) -> (args: [String], out: URL) {
        let out = outputURL(for: input, suffix: "-mp4", ext: "mp4")
        return (["-i", input.path, "-c:v", "libx264", "-preset", "veryfast",
                 "-c:a", "aac", out.path], out)
    }

    /// Trim a clip. Re-encodes (rather than `-c copy`) so the cut is frame-
    /// accurate across containers; `veryfast` keeps it quick.
    static func trim(_ input: URL, start: String, end: String) -> (args: [String], out: URL) {
        let out = outputURL(for: input, suffix: "-clip", ext: "mp4")
        return (["-ss", start, "-to", end, "-i", input.path,
                 "-c:v", "libx264", "-preset", "veryfast", "-c:a", "aac", out.path], out)
    }
}

File: windows/tapenexus/media_tools.py
"""Post-download tools that run the bundled ffmpeg.exe on a Library file:
extract audio, transcode to MP4, and trim a clip. ffmpeg is the same binary
yt-dlp uses for merging, so no new dependency is shipped."""
from __future__ import annotations

import os
from typing import List, Tuple


def _output_url(inp: str, suffix: str, ext: str) -> str:
    """A non-clobbering output path beside the input: <name><suffix>.<ext>."""
    d = os.path.dirname(inp)
    base = os.path.splitext(os.path.basename(inp))[0]
    candidate = os.path.join(d, f"{base}{suffix}.{ext}")
    i = 2
    while os.path.exists(candidate):
        candidate = os.path.join(d, f"{base}{suffix}-{i}.{ext}")
        i += 1
    return candidate


def extract_audio_mp3(inp: str) -> Tuple[List[str], str]:
    out = _output_url(inp, "-audio", "mp3")
    return (["-i", inp, "-vn", "-c:a", "libmp3lame", "-q:a", "2", out], out)


def extract_audio_aac(inp: str) -> Tuple[List[str], str]:
    out = _output_url(inp, "-audio", "m4a")
    return (["-i", inp, "-vn", "-c:a", "aac", "-b:a", "192k", out], out)


def transcode_mp4(inp: str) -> Tuple[List[str], str]:
    out = _output_url(inp, "-mp4", "mp4")
    return (["-i", inp, "-c:v", "libx264", "-preset", "veryfast", "-c:a", "aac", out], out)


def trim(inp: str, start: str, end: str) -> Tuple[List[str], str]:
    """Trim a clip. Re-encodes for a frame-accurate cut across containers."""
    out = _output_url(inp, "-clip", "mp4")
    return (["-ss", start, "-to", end, "-i", inp,
             "-c:v", "libx264", "-preset", "veryfast", "-c:a", "aac", out], out)